### PR TITLE
qa: fill the test_file before trying to read it

### PR DESCRIPTION
**Pull request by lxbsz**
When the file is empty the kernel will trigger a getattr request
to libcephfs to get the file size, and then if the read offset is
larger or equal to the file size it will just return without firing
real read to libcephfs.

This will fill the file with some contents and reopen the file to
flush the buffer and reset the file offset to 0.

More detail please see:
https://tracker.ceph.com/issues/53859?issue_count=49&issue_position=2&next_issue_id=53908&prev_issue_id=43960#note-5

Fixes: https://tracker.ceph.com/issues/53859
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by batrick — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_pool_perm.py [hunk 1]
```diff
@@ -17,13 +17,18 @@ def test_pool_perm(self):
...
+                    fd = os.open("{path}", os.O_RDWR)
                     ret = os.read(fd, 1024)
                 else:
                     os.write(fd, b'content')
             except OSError as e:
                 if e.errno != errno.EPERM:
+                    os.close(fd)
                     raise
             else:
+                os.close(fd)
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply qa/tasks/cephfs/test_pool_perm.py]: Done.

**Review by vshankar — APPROVED** [review]

**Comment by lxbsz** [comment]
jenkins test make check

**Comment by vshankar** [comment]
@lxbsz This test still fails - https://pulpito.ceph.com/vshankar-2022-02-05_17:27:49-fs-wip-vshankar-testing-20220201-113815-testing-default-smithi/6662928/ (build https://shaman.ceph.com/builds/ceph/wip-vshankar-testing-20220201-113815/)

PTAL.

**Comment by lxbsz** [comment]
> @lxbsz This test still fails - https://pulpito.ceph.com/vshankar-2022-02-05_17:27:49-fs-wip-vshankar-testing-20220201-113815-testing-default-smithi/6662928/ (build https://shaman.ceph.com/builds/ceph/wip-vshankar-testing-20220201-113815/)
> 
> PTAL.

This time it fails in the test for kclient. I will check how the kclient is handling this case.

**Comment by lxbsz** [comment]
> > @lxbsz This test still fails - https://pulpito.ceph.com/vshankar-2022-02-05_17:27:49-fs-wip-vshankar-testing-20220201-113815-testing-default-smithi/6662928/ (build https://shaman.ceph.com/builds/ceph/wip-vshankar-testing-20220201-113815/)
> > PTAL.
> 
> This time it fails in the test for kclient. I will check how the kclient is handling this case.

This time it failed when testing the writing to the file and there has no logs, but from the code it should due to buffer writing and just successfully wrote to the buffer.

**Comment by vshankar** [comment]
> This time it failed when testing the writing to the file and there has no logs, but from the code it should due to buffer writing and just successfully wrote to the buffer.



> This time it failed when testing the writing to the file and there has no logs, but from the code it should due to buffer writing and just successfully wrote to the buffer.

OK -- O_DIRECT should help here.

I'll include this in my next run. Thanks @lxbsz

**Comment by vshankar** [comment]
@lxbsz - please take a look - https://pulpito.ceph.com/vshankar-2022-03-06_14:47:51-fs-wip-vshankar-testing-20220304-132102-testing-default-smithi/6724272/

(I didn't check why it fails with your latest changes - sorry!)

**Comment by lxbsz** [comment]
> @lxbsz - please take a look - https://pulpito.ceph.com/vshankar-2022-03-06_14:47:51-fs-wip-vshankar-testing-20220304-132102-testing-default-smithi/6724272/
> 
> (I didn't check why it fails with your latest changes - sorry!)

No worry, I will check it later. Thanks Venky.

**Comment by lxbsz** [comment]
Reopen it and need to update this PR to find the root cause later.

**Review by rishabh-d-dave — APPROVED** [review]
Looks okay to me.

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
jenkins test make windows

**Comment by rishabh-d-dave** [comment]
jenkins test docs

**Comment by rishabh-d-dave** [comment]
jenkins test windows

**Comment by rishabh-d-dave** [comment]
jenkins test make check arm64

**Comment by rishabh-d-dave** [comment]
@lxbsz @vshankar Shall I pick this for QA?

**Comment by vshankar** [comment]
> @lxbsz @vshankar Shall I pick this for QA?

Yes.

**Comment by lxbsz** [comment]
> @rishabh-d-dave sorry, I think we should wait for @lxbsz to push a change update.

Yeah, last time this test failed and I need to spend more time reading the kernel VFS code to found the root cause.

**Comment by rishabh-d-dave** [comment]
@vshankar @lxbsz Okay, I'll wait. Please feel free me to message me when it's ready.

**Comment by rishabh-d-dave** [comment]
Hey, @lxbsz. This PR got autoclosed. Just checking if you too want it to remain closed or you want to re-open it. :)

**Comment by lxbsz** [comment]
> Hey, @lxbsz. This PR was autoclosed. Just checking if you too want it to remain closed or you want to re-open it. :)

Thanks @rishabh-d-dave I just opened it, and this needs to be updated later.

**Comment by rishabh-d-dave** [comment]
> > Hey, @lxbsz. This PR was autoclosed. Just checking if you too want it to remain closed or you want to re-open it. :)
> 
> Thanks @rishabh-d-dave I just opened it, and this needs to be updated later.

Just FYI: This PR got auto-closed again.

**Pull request closed without merging**